packages/stylist-core/src/ast/context.rs

```diff
@@ -0,0 +1,174 @@
+use std::borrow::Cow;
+use std::sync::atomic::{AtomicBool, Ordering};
+
+/// A context to faciliate [`ToStyleStr`](super::ToStyleStr).
+#[derive(Debug)]
+pub struct StyleContext<'a> {
+    pub class_name: Option<&'a str>,
+    parent_ctx: Option<&'a StyleContext<'a>>,
+
+    rules: Vec<Cow<'a, str>>,
+    selectors: Vec<Cow<'a, str>>,
+
+    is_open: AtomicBool,
```

Comment:
This is due to mutability restraint.

If `is_open` is `bool`, then `StyleContext` needs to hold a mutable reference of `parent_ctx` in order to instruct `finish()` on parents, in which I cannot get the compiler to satisfy.

Reply:
Ah yeah, since mutable references are invariant in the type they point to, this can mess with some coercions that happen automatically with an immutable one, which is covariant in that type, which seems to happen when you reborrow in the recursive calls in `write_style`.

In any case, this is cleaner than introducing weird lifetimes or other workaround, for now.